core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -375,24 +385,54 @@ class LogManager(val logDirs: Array[File],
   }
 
   /**
-   *  Delete a log.
+   *  Delete logs marked for deletion.
    */
-  def deleteLog(topicAndPartition: TopicAndPartition) {
-    var removedLog: Log = null
-    logCreationOrDeletionLock synchronized {
-      removedLog = logs.remove(topicAndPartition)
+  private def deleteLogs(): Unit = {
+    while (!logsToBeDeleted.isEmpty) {
+      val removedLog = logsToBeDeleted.take()
```

Comment:
If there are any storage exceptions in delete, we probably want to retry - i.e., we should `peek` and `take` only after the delete succeeds.